### PR TITLE
Compile form extensions with node

### PR DESCRIPTION
**Pull request by tamarasaurus**
**Description (for Contributor and Core Developer)**

This PR separates the form extension parsing from the backend. Collecting and merging all the form extension files from each bundle is now handled in node and executed before each webpack build. 

Before
1. PHP dumps extensions.json which contains the form extensions from each bundle
2. The frontend calls `/form/extensions` to get the extensions filtered by ACL and loads the app

The old way had caching at two points, for the yml files and also in the browser for the request. The new way:

1. Node dumps the extensions.json
2. The frontend requests the dumped file `/js/extensions.json`, filters ACL on the frontend (using security-context) and loads the app

So with the new way there is no need to clear the symfony cache each time a form extension yml file is changed. Webpack will update the extensions.json on each build since it only takes around half a second. 

Another improvement is that if you are running `yarn webpack-watch`, webpack will listen for changes in the form extension yml files, then update the extensions.json and build the app automatically. 

To update the extensions manually you can also run `yarn update-extensions` (if you didn't change any JS). This replaces the command `bin/console pim:installer:dump-extensions`

Changes
- [x] Compile extensions in node
- [x] Remove form extensions route from backend
- [x] Remove dump extensions command from backend
- [x] Add `yarn update-extensions` command and execute it on webpack build
- [x] Add automatic rebuild in webpack-watch for extensions
- [x] Simplify webpack config loader
- [x] EE Standard PR - https://github.com/akeneo/pim-enterprise-standard/pull/768
- [x] CE Standard PR - https://github.com/akeneo/pim-community-standard/pull/883
- [ ] Docs PR

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | -
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | -
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: frontend/build/compile-less.js [hunk 1]
```diff
@@ -12,14 +12,13 @@ const RewriteImageURLs = require('./less-rewrite-urls')
...
-const BUNDLE_LESS_INDEX_PATH = 'public/less/index.less'
+const BUNDLE_LESS_INDEX_PATH = 'Resources/public/less/index.less'
 
 // The final output path for all the CSS of the PIM
 const OUTPUT_CSS_PATH = 'web/css/pim.css'
 
-if (!existsSync(`${BUNDLE_REQUIRE_PATH}.js`)) {
-    console.error(`web/js/require-paths.js does not exist - Run "bin/console pim:installer:dump-require-paths" and try again.`.red)
-
+if (!existsSync(BUNDLE_REQUIRE_PATH + '.js')) {
```
Comment (juliensnz) [inline frontend/build/compile-less.js]: just to understand, why are you not using string templates?

File: frontend/build/update-extensions.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+function mergeExtensions(paths) {
+    const config = paths.map(path => getFileContents(path))
+    const merged = deepmerge.all(config)
+    const configuredExtensions = {}
+    let i = 0;
+
+    for (let extension in merged.extensions) {
+        const extensionConfig = _.defaults(
```
Comment (juliensnz) [inline frontend/build/update-extensions.js]: why not using simple [...merged.extensions[extension], ... EXTENSION_DEFAULTS]?

File: frontend/build/update-extensions.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    return [].concat.apply([], formExtensions);
+}
+
+function mergeExtensions(paths) {
+    const config = paths.map(path => getFileContents(path))
+    const merged = deepmerge.all(config)
+    const configuredExtensions = {}
+    let i = 0;
+
+    for (let extension in merged.extensions) {
```
Comment (juliensnz) [inline frontend/build/update-extensions.js]: why not an array map?

File: frontend/build/update-extensions.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
+require('colors')
+const rootDir = process.cwd()
+const { resolve, dirname } = require('path')
+const glob = require('glob')
+const { parse } = require('yamljs')
+const { readFileSync, writeFileSync } = require('fs')
+const deepmerge = require('deepmerge')
+const _ = require('lodash');
```
Comment (juliensnz) [inline frontend/build/update-extensions.js]: You could remove this dependency using only es6/7 feature? Is it a deliberated choice?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/config-provider.js [hunk 5]
```diff
@@ -1,16 +1,37 @@
...
             if (null === promise) {
-                promise = $.getJSON(Routing.generate('pim_enrich_form_extension_rest_index')).fail(() => {
-                    throw Error('It seems that your web server is not well configured as we were not able ' +
-                        'to load the frontend configuration. The most likely reason is that the mod_rewrite ' +
-                        'module is not installed/enabled.')
+                promise = $.when(
+                    $.get('/js/extensions.json'),
+                    SecurityContext.initialize()
+                ).then((formExtensions) => {
+                    const test = formExtensions[0]
```
Comment (tamarasaurus) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/config-provider.js]: rename

Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/public/js/form/config-provider.js]: test?

File: frontend/build/compile-less.js (hunk 1 above)
Comment (tamarasaurus) [reply frontend/build/compile-less.js]: no reason, just a mistake

File: frontend/build/compile-less.js [hunk 6]
```diff
@@ -12,14 +12,13 @@ const RewriteImageURLs = require('./less-rewrite-urls')
...
-const BUNDLE_LESS_INDEX_PATH = 'public/less/index.less'
+const BUNDLE_LESS_INDEX_PATH = 'Resources/public/less/index.less'
 
 // The final output path for all the CSS of the PIM
 const OUTPUT_CSS_PATH = 'web/css/pim.css'
 
 if (!existsSync(`${BUNDLE_REQUIRE_PATH}.js`)) {
-    console.error(`web/js/require-paths.js does not exist - Run "bin/console pim:installer:dump-require-paths" and try again.`.red)
-
+    console.log(`${BUNDLE_REQUIRE_PATH} does not exist - Run "bin/console pim:installer:dump-require-paths" and try again.`.red)
```
Comment (phaseinducer) [inline frontend/build/compile-less.js]: You already did this in `check-requirements.js`, do you want to check it again ?

File: frontend/build/update-extensions.js [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+ *
+ * Returns an object containing the attribute fields and form extensions sorted by position
+ *
+ * @param {string} paths
+ */
+function mergeExtensions(paths) {
+    const config = paths.map(path => getFileContents(path))
+    const merged = deepmerge.all(config)
+    const mergedExtensions = Object.entries(merged.extensions).map(([code, extension]) => {
+        return extensionConfig = {
```
Comment (phaseinducer) [inline frontend/build/update-extensions.js]: Why not just returning directly ?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/attribute/file.yml [hunk 8]
```diff
@@ -59,6 +59,6 @@ extensions:
         config:
             fieldName: allowed_extensions
             label: pim_enrich.entity.attribute.property.allowed_extensions
-            choices: '%pim_catalog_file_allowed_extensions%'
+            choices: [csv, doc, docx, mp3, pdf, ppt, pptx, rtf, svg, txt, wav]
```
Comment (phaseinducer) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/attribute/file.yml]: Why changing that ?

File: frontend/build/compile-less.js (hunk 6 above)
Comment (tamarasaurus) [reply frontend/build/compile-less.js]: check-requirements doesn't run before .less compilation

File: frontend/build/update-extensions.js (hunk 7 above)
Comment (tamarasaurus) [reply frontend/build/update-extensions.js]: good point

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/attribute/file.yml (hunk 8 above)
Comment (tamarasaurus) [reply src/Akeneo/Platform/Bundle/UIBundle/Resources/config/form_extensions/attribute/file.yml]: The backend was populating this field, after my changes that's not possible anymore. I'll update the documentation to reflect that

**Pull request merged**